internal/extproc/embeddings_processor.go

```diff
@@ -0,0 +1,393 @@
+	}()
+	var br io.Reader
+	var isGzip bool
+	switch e.responseEncoding {
+	case "gzip":
+		br, err = gzip.NewReader(bytes.NewReader(body.Body))
+		if err != nil {
+			return nil, fmt.Errorf("failed to decode gzip: %w", err)
+		}
+		isGzip = true
+	default:
+		br = bytes.NewReader(body.Body)
+	}
+
+	headerMutation, bodyMutation, tokenUsage, err := e.translator.ResponseBody(e.responseHeaders, br, body.EndOfStream)
+	if err != nil {
+		return nil, fmt.Errorf("failed to transform response: %w", err)
+	}
+	if bodyMutation != nil && isGzip {
+		if headerMutation == nil {
+			headerMutation = &extprocv3.HeaderMutation{}
+		}
+		// TODO: this is a hotfix, we should update this to recompress since its in the header
+		// If the response was gzipped, ensure we remove the content-encoding header.
+		//
+		// This is only needed when the transformation is actually modifying the body. When the backend
+		// is in OpenAI format (and it's the first try before any retry), the response body is not modified,
+		// so we don't need to remove the header in that case.
+		headerMutation.RemoveHeaders = append(headerMutation.RemoveHeaders, "content-encoding")
+	}
+
+	resp := &extprocv3.ProcessingResponse{
+		Response: &extprocv3.ProcessingResponse_ResponseBody{
+			ResponseBody: &extprocv3.BodyResponse{
+				Response: &extprocv3.CommonResponse{
+					HeaderMutation: headerMutation,
+					BodyMutation:   bodyMutation,
+				},
+			},
+		},
+	}
+
+	// Accumulate token usage for embeddings (only input and total tokens are relevant)
+	e.costs.InputTokens += tokenUsage.InputTokens
+	e.costs.TotalTokens += tokenUsage.TotalTokens
+
+	// Update metrics with token usage.
+	e.metrics.RecordTokenUsage(ctx, tokenUsage.InputTokens, tokenUsage.TotalTokens)
+
+	if body.EndOfStream && len(e.config.requestCosts) > 0 {
+		resp.DynamicMetadata, err = e.maybeBuildDynamicMetadata()
+		if err != nil {
+			return nil, fmt.Errorf("failed to build dynamic metadata: %w", err)
+		}
+	}
+
+	return resp, nil
+}
+
+// SetBackend implements [Processor.SetBackend].
+func (e *embeddingsProcessorUpstreamFilter) SetBackend(ctx context.Context, b *filterapi.Backend, backendHandler backendauth.Handler, routeProcessor Processor) (err error) {
+	defer func() {
+		e.metrics.RecordRequestCompletion(ctx, err == nil)
+	}()
+	rp, ok := routeProcessor.(*embeddingsProcessorRouterFilter)
+	if !ok {
+		panic("BUG: expected routeProcessor to be of type *embeddingsProcessorRouterFilter")
+	}
+	rp.upstreamFilterCount++
+	e.metrics.SetBackend(b)
+	e.modelNameOverride = b.ModelNameOverride
+	if err = e.selectTranslator(b.Schema); err != nil {
+		return fmt.Errorf("failed to select translator: %w", err)
+	}
+	e.handler = backendHandler
+	e.originalRequestBody = rp.originalRequestBody
+	e.originalRequestBodyRaw = rp.originalRequestBodyRaw
+	e.onRetry = rp.upstreamFilterCount > 1
+	rp.upstreamFilter = e
+	return
+}
+
+func (e *embeddingsProcessorUpstreamFilter) maybeBuildDynamicMetadata() (*structpb.Struct, error) {
```

Comment:
can you generalize this and share it with chat completion?

Reply:
Working on it